### PR TITLE
SPLAT-1452: Added New FeatureGate/FeatureSet Aware MaxItems

### PR DESCRIPTION
**Pull request by vr4manta**
[SPLAT-1452](https://issues.redhat.com/browse/SPLAT-1452)

### Changes
- Added new FeatureSetAwareMaxItems
- Added new FeatureGateAwareMaxItems

### Blocks
- https://github.com/openshift/api/pull/1842

### Use Case
This enhancement is being done for the vSphere Multi vCenter feature. The idea is we want to have max vCenter count == 3 initially for Tech Preview. By adding maxItemsAware we will have a way to set Default stay at 1, but for TP / Custom, allow vCenter count to be a max of 3.  Also by setting this way with the maxItemsAware, we will be able to have the CRD limit it so all operators do not have to be max count aware (check for 3 vs just checking for feature enabled).

Example of new validation field:
> // +openshift:validation:FeatureGateAwareMaxItems:featureGate="",maxItems=1
> // +openshift:validation:FeatureGateAwareMaxItems:featureGate=VSphereMultiVCenters,maxItems=3

### CODE REVIEW

File: pkg/crd/markers/patch_validation.go [hunk 1]
```diff
@@ -143,6 +149,25 @@ func (m FeatureGateEnum) ApplyToSchema(schema *apiext.JSONSchemaProps) error {
...
+	MaxItems           int      `marker:"maxItems"`
+}
+
+func (m FeatureGateMaxItems) ApplyToSchema(schema *apiext.JSONSchemaProps) error {
+	if FeatureGatesForCurrentFile.HasAny(m.FeatureGateNames...) != m.FeatureGateEnabled {
+		return nil
+	}
+
+	if schema.Type != "array" {
+		return fmt.Errorf("must apply maxitem to an array")
```
Comment (JoelSpeed) [inline pkg/crd/markers/patch_validation.go:163]: Is it `maxitem` cased in other error messages?

File: pkg/crd/markers/patch_validation.go [hunk 2]
```diff
@@ -143,6 +149,25 @@ func (m FeatureGateEnum) ApplyToSchema(schema *apiext.JSONSchemaProps) error {
 	return nil
 }
 
+type FeatureGateMaxItems struct {
+	FeatureGateNames   []string `marker:"featureGate"`
+	FeatureGateEnabled bool     `marker:"featureGateEnabled"`
```
Comment (JoelSpeed) [inline pkg/crd/markers/patch_validation.go]: What's this looking at? Can you provide an example of using this?

I guess this is so you can have one version with the feature gate and one without? Is this how it works for other feature gate aware markers?

Comment (vr4manta) [reply pkg/crd/markers/patch_validation.go]: The featureGateEnabled was a test i tried.  I can remove that one to match the others.  For some reason I thought I had copied that from another existing one, but now I am not sure.

Comment (vr4manta) [reply pkg/crd/markers/patch_validation.go]: Oh I remember now, I needed a way to say if my feature gate is disabled, to do something.  See example in comment about syntax above.  Also, the PR from blocker section in PR description has usage and outputs to show this logic working.

File: pkg/crd/markers/patch_validation.go (hunk 1 above)
Comment (vr4manta) [reply pkg/crd/markers/patch_validation.go:163]: I copied this logic from here: https://github.com/openshift/kubernetes-sigs-controller-tools/blob/62ad93abdfc613b59248f06d415a006d9f2d8bb4/pkg/crd/markers/validation.go#L376

They are using the all lowercase for their output.  I can change to maxItems to match the actual field name.

Examples of FeatureGateAwere and normal:

Normal:
> // +kubebuilder:validation:MaxItems=20

FG Aware
> 	// +openshift:validation:FeatureGateAwareMaxItems:featureGate=VSphereMultiVCenters,featureGateEnabled=false,maxItems=1
>	// +openshift:validation:FeatureGateAwareMaxItems:featureGate=VSphereMultiVCenters,featureGateEnabled=true,maxItems=3

The base has var start w/ capital, but feature gate aware starts with lower and upper depending.  It seems its not case sensitive.

Example:
> // +openshift:validation:FeatureGateAwareEnum:featureGate="",enum="";None;IntegratedOAuth

or

> // +openshift:validation:FeatureGateAwareXValidation:featureGate=GCPLabelsTags,rule="!has(oldSelf.resourceLabels) && !has(self.resourceLabels) || has(oldSelf.resourceLabels) && has(self.resourceLabels)",message="resourceLabels may only be configured during installation"
> // +openshift:validation:FeatureGateAwareXValidation:featureGate=GCPLabelsTags,rule="!has(oldSelf.resourceTags) && !has(self.resourceTags) || has(oldSelf.resourceTags) && has(self.resourceTags)",message="resourceTags may only be configured during installation"

Comment (JoelSpeed) [reply pkg/crd/markers/patch_validation.go:163]: If you look at [how the enum](https://github.com/openshift/api/blob/8203151f085f2205ba7fa64f093a04a742d1cb87/example/v1/types_stable.go#L79-L80) version of this works, the default case is with an empty feature gate, can we copy the same pattern so there's consistency? 

I'd also like an example usage of this marker added to the same example APi once this is merged please

Comment (vr4manta) [reply pkg/crd/markers/patch_validation.go:163]: So problem with default when I tested it was that it was also firing casuing my value of 3 to get overridden by the 1.  Thats why I added the second flag.  I can retest this , but I am pretty sure this is what I saw resulting in me adding this new variable.

For the example api, I can add it there.  I haven't seen the example before so I will ping you offline about this.  Thanks!

Comment (vr4manta) [reply pkg/crd/markers/patch_validation.go:163]: Based on our discussion, I've made the changes to follow enum's example.  I've confirmed this is looking great.  I have pushed the changes.

For the api example, I'll make that change and push w/ my PR at top of this PR (marked as blocker).

**Comment by JoelSpeed** [comment]
Spoken with @vr4manta, this now is in keeping with the enum version for usage, an example will be added to the o/api examples as well

**Pull request merged**